mirror/mirror.go

```diff
@@ -290,43 +342,191 @@ func (h *Handle) fetchBatch(number uint32) error {
 
 	if err := getAndStore(
 		h.url(prefix+"evidence"),
-		evPath,
+		evsPath,
 	); err != nil {
 		return err
 	}
 
-	// TODO umbilical certificates
+	var (
+		ucs      []*frozencas.Handle
+		umbRoots *x509.CertPool
+	)
+	if h.b.Params.EvidencePolicy == mtc.UmbilicalEvidencePolicy {
+		umbRoots, err = h.b.GetUmbilicalRoots()
+		if err != nil {
+			return err
+		}
+
+		if err := getAndStore(
+			h.url(prefix+"umbilical-certificates"),
+			ucPath,
+		); err != nil {
+			return err
+		}
+
+		// Oldest batch to inspect for deduplicated umbilical certificate
+		end := int64(batch.Number) - int64(h.b.Params.ValidityWindowSize)
+		if end < 0 {
+			end = 0
+		}
+
+		for bn := int64(batch.Number) - 1; bn >= end; bn-- {
+			uc, err := h.b.UCFor(uint32(bn))
+			if err != nil {
+				return fmt.Errorf(
+					"opening umbilical certificates for batch %d: %w",
+					bn,
+					err,
+				)
+			}
+
+			ucs = append(ucs, uc)
+		}
+	}
 
 	var svw mtc.SignedValidityWindow
 	if err := svw.UnmarshalBinary(svwBuf, &h.b.Params); err != nil {
 		return fmt.Errorf("parsing validity-window: %w", err)
 	}
 
-	// Recompute tree
+	// Prepare to recompute tree and check evidence
 	tb := batch.NewTreeBuilder()
+
 	besR, err := os.OpenFile(besPath, os.O_RDONLY, 0)
 	if err != nil {
 		return fmt.Errorf("opening %s: %w", besPath, err)
 	}
 	defer besR.Close()
 
+	evsR, err := os.OpenFile(evsPath, os.O_RDONLY, 0)
+	if err != nil {
+		return fmt.Errorf("opening %s: %w", evsPath, err)
+	}
+	defer evsR.Close()
+
 	besC := mtc.UnmarshalBatchEntries(bufio.NewReader(besR))
 	defer besC.Close()
 
-	var be mtc.BatchEntry
+	evsC := mtc.UnmarshalEvidenceLists(bufio.NewReader(evsR))
+	defer evsC.Close()
+
+	var (
+		be  mtc.BatchEntry
+		evs mtc.EvidenceList
+	)
+
+	entryNumber := -1
 	for {
```

Comment:
Might be more natural to write this as `for entryNumber := 0; ; entryNumber++ { ...`

Reply:
Oh, of course!